### PR TITLE
Fix exec_replace

### LINKED ISSUE #125: Properly implement `scarb::process::exec_replace`
Fix the TODO comment left there. Doc comment should fully describe what should happen internally.

https://github.com/software-mansion/scarb/blob/7bf281cc01ddc9791275a57aa91f65077b7795f4/scarb/src/process.rs#L14-L48

### LINKED ISSUE #129: Fix subcommand CTRL+C test
https://github.com/software-mansion/scarb/blob/7bf281cc01ddc9791275a57aa91f65077b7795f4/scarb/tests/e2e/subcommand.rs#L140-L182

### PR DESCRIPTION
**Pull request by MrDenkoV**
Fix #125, Fix #129

### CODE REVIEW

File: scarb/tests/subcommand.rs [hunk 1]
```diff
@@ -115,9 +115,7 @@ fn env_scarb_log_is_passed_verbatim() {
         .success();
 }
 
-// TODO(#129): Fix this test.
 #[test]
-#[ignore] // something doesn't work here
```
Comment (mkaput) [inline scarb/tests/subcommand.rs:120]: WOW

File: scarb/src/process.rs
Comment (mkaput) [inline scarb/src/process.rs:1]: can you please remove the TODO line here?: https://github.com/software-mansion/scarb/pull/658/files#diff-c6dd60aae938af440d1b0c7787bdcdd9330cd65cb1c6d5ebec2e3209f7ac8791R15

File: scarb/src/process.rs [hunk 2]
```diff
@@ -30,21 +30,58 @@ use crate::core::Config;
...
+
+    unsafe extern "system" fn ctrlc_handler(_: u32) -> BOOL {
+        // Do nothing; let the child process handle it.
+        TRUE
+    }
+
+    pub fn exec_replace(cmd: &Command) -> Result<()> {
+        unsafe {
+            if SetConsoleCtrlHandler(Some(ctrlc_handler), TRUE) == FALSE {
+                return Err(ProcessError::new("Could not set Ctrl-C handler.", None, None).into());
```
Comment (mkaput) [inline scarb/src/process.rs:63]: 1. Where does `ProcessError` come from?
2. Woudln't it make sense to just panic here?

File: scarb/src/process.rs [hunk 3]
```diff
@@ -30,21 +30,58 @@ use crate::core::Config;
...
+    use std::process::Command;
+
+    pub fn exec_replace(cmd: &mut Command) -> Result<()> {
+        let err = cmd.exec();
+        bail!("process did not exit successfully: {err}")
+    }
+}
+
+#[cfg(windows)]
+mod imp {
```
Comment (mkaput) [inline scarb/src/process.rs:49]: Have you tested this on Windows?

File: scarb/src/process.rs [hunk 4]
```diff
@@ -30,21 +30,58 @@ use crate::core::Config;
...
-
-    if exit_status.success() {
-        Ok(())
-    } else {
-        bail!("process did not exit successfully: {exit_status}");
+    imp::exec_replace(cmd)
+}
+
+#[cfg(unix)]
+mod imp {
```
Comment (mkaput) [inline scarb/src/process.rs:37]: Does this compile and run on Linux and macOS?

**Review by mkaput — CHANGES_REQUESTED** [review]
I am amazed that this fixes #129 

Let's test this more thoroughly

File: scarb/src/process.rs (hunk 2 above)
Comment (MrDenkoV) [reply scarb/src/process.rs:63]: 😅

File: scarb/src/process.rs (hunk 3 above)
Comment (MrDenkoV) [reply scarb/src/process.rs:48]: Doesn't work on Windows, our test hangs... 😰

**Comment by maciektr** [comment]
👋 
On the topic of our offline conversation about ctrl_c test on windows 🪟 :

The error you have been seeing seems to come from the subcommand spawn. 
```
This version of %1 is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher. (os error 216)
```
It seems that the script definition used there is just not runnable on windows. (https://github.com/software-mansion/scarb/blob/f8494ed3290c5ae45aa7015a2cf36f38c12f58ff/scarb/tests/subcommand.rs#L144) 

After changing the script definition, I was able to run it successfully and ensure it has been killed by ctrl_c. 
To test this, I have created a `main.rs` file in `scarb-test-support` crate with following content: https://gist.github.com/maciektr/6b6c18a358b5792912bc0d0261ef3ac2
Then I have used this in tests instead.
```
    let addr = {
        let addr = listener.local_addr().unwrap();
        addr.to_string()
    };

    let mut child = Scarb::new()
        .std()
        .arg("test-support")
        .arg("hang-on-tcp")
        .arg("--address")
        .arg(addr)
```


I have not been able to make this check work though. I do not understand what is the purpose of this. 
```
    match sock.read(&mut [0; 10]) {
        Ok(n) => assert_eq!(n, 0),
        Err(e) => assert_eq!(e.kind(), io::ErrorKind::ConnectionReset),
    }
```
I know it has been copied from Cargo verbatim. Maybe it is enough for now to only check if the process has been killed? cc @mkaput

**Comment by mkaput** [comment]
> I have not been able to make this check work though. I do not understand what is the purpose of this.
> 
> ```
>     match sock.read(&mut [0; 10]) {
>         Ok(n) => assert_eq!(n, 0),
>         Err(e) => assert_eq!(e.kind(), io::ErrorKind::ConnectionReset),
>     }
> ```
> 
> I know it has been copied from Cargo verbatim. Maybe it is enough for now to only check if the process has been killed? cc @mkaput

This check tests if the socket either closed successfully without any unread rubbish left in the buffer, or died by the second party being terminated.

> It seems that the script definition used there is just not runnable on windows. (
> 
> https://github.com/software-mansion/scarb/blob/f8494ed3290c5ae45aa7015a2cf36f38c12f58ff/scarb/tests/subcommand.rs#L144
> 
> )
> After changing the script definition, I was able to run it successfully and ensure it has been killed by ctrl_c. To test this, I have created a `main.rs` file in `scarb-test-support` crate with following content: https://gist.github.com/maciektr/6b6c18a358b5792912bc0d0261ef3ac2 Then I have used this in tests instead.
> 
> ```
>     let addr = {
>         let addr = listener.local_addr().unwrap();
>         addr.to_string()
>     };
> 
>     let mut child = Scarb::new()
>         .std()
>         .arg("test-support")
>         .arg("hang-on-tcp")
>         .arg("--address")
>         .arg(addr)
> ```
> 
> May not be the cleanest flow, but imho worth considering.

I like this! I'd only rather use `Command::new(cargo_bin!("scarb-test-support"))` instead of `Scarb::new().std()`, seems more clear.

File: scarb/src/process.rs (hunk 3 above)
Comment (mkaput) [reply scarb/src/process.rs:48]: How is it now?

File: utils/scarb-test-support/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+fn hang_on_tcp(args: HangOnTcpArgs) -> Result<()> {
+    use std::io::Read;
+    use std::net::TcpStream;
+
+    let address: &str = args.address.as_ref();
+
+    let mut socket = TcpStream::connect(address).unwrap();
+    let _ = socket.read(&mut [0; 10]);
+    panic!("that read should never return");
```
Comment (mkaput) [inline utils/scarb-test-support/src/main.rs:38]: https://doc.rust-lang.org/stable/std/macro.unreachable.html

File: utils/scarb-test-support/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    address: String,
+}
+
+fn main() {
+    let args: Args = Args::parse();
+    let result = match args.command {
+        Command::HangOnTcp(args) => hang_on_tcp(args),
+    };
+    result.unwrap();
+}
```
Comment (mkaput) [inline utils/scarb-test-support/src/main.rs:28]:
```suggestion
fn main() -> Result<()> {
    let args = Args::parse();
    match args.command {
        Command::HangOnTcp(args) => hang_on_tcp(args),
    }
}
```

File: scarb/tests/subcommand.rs [hunk 7]
```diff
@@ -115,38 +117,18 @@ fn env_scarb_log_is_passed_verbatim() {
...
 fn ctrl_c_kills_everyone() {
-    let t = assert_fs::TempDir::new().unwrap();
     let listener = TcpListener::bind("127.0.0.1:0").unwrap();
+    let addr = {
+        let addr = listener.local_addr().unwrap();
+        addr.to_string()
+    };
```
Comment (mkaput) [inline scarb/tests/subcommand.rs:126]:
```suggestion
    let addr = listener.local_addr().unwrap().to_string();
```

File: scarb/src/process.rs (hunk 3 above)
Comment (MrDenkoV) [reply scarb/src/process.rs:48]: It works! C:

File: utils/scarb-test-support/tests/examples.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+use scarb_test_support::command::Scarb;
+
+#[test]
+fn test_help() {
```
Comment (mkaput) [inline utils/scarb-test-support/tests/examples.rs:4]: what's the reason behind adding this?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: utils/scarb-test-support/tests/examples.rs (hunk 8 above)
Comment (MrDenkoV) [reply utils/scarb-test-support/tests/examples.rs:4]: It's for cargo test to create the binary "scarb-test-support".
It helps cargo_bin("scarb-test-support").

Comment (mkaput) [reply utils/scarb-test-support/tests/examples.rs:4]: Can you please:
1. Add an explanatory comment to this file.
2. Rename the test `binary_dependencies_hack` and leave its body empty. I do not want to test help output, especially in such exotic place.
3. Rename test file to something more meaningful. Even simple `hack.rs` is better than `examples.rs` IMO

**Review by mkaput — CHANGES_REQUESTED** [review]

File: utils/scarb-test-support/tests/examples.rs (hunk 8 above)
Comment (mkaput) [reply utils/scarb-test-support/tests/examples.rs:4]: And also, please create an issue in our tracker to fix this hack with https://github.com/rust-lang/cargo/issues/9096 once it's released in stable Rust.

Comment (MrDenkoV) [reply utils/scarb-test-support/tests/examples.rs]: Done!

**Review by mkaput — APPROVED** [review]

**Pull request merged**